Improve wp_exif_frac2dec

adamsilverstein opened the pull request:
Fix an issue where certain images could cause a fatal error under PHP 8+, because `wp_exif_frac2dec` unexpectedly returns a string for certain images.

Since PHP 8, argument #1 ($num) passed to the function `round()` must be of type int|float, otherwise a `PHP Fatal error:  Uncaught TypeError` error is thrown.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/54385
<!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
Cast value return from wp_exif_frac2dec to float before rounding.

Fix an issue where certain images could cause a fatal error under PHP 8+, because `wp_exif_frac2dec` unexpectedly returns a string for certain images.

Since PHP 8, argument #1 ($num) passed to the function `round()` must be of type int|float, otherwise a `PHP Fatal error:  Uncaught TypeError` error is thrown.

adamsilverstein committed:
update doc block

adamsilverstein committed:
Ensure wp_exif_frac2dec returns float or it.

jrfnl reviewed (commented):
@adamsilverstein 👋🏻 Hope life is treating you well!

Sorry to butt in... If I look at the function description and what the function actually _does_, I'm a little confounded...

There is no check whatsoever that the actual input _could_ be a fraction, i.e. consists of numbers divided by a '/'.

Just as an example: someone could pass it a path string `path/to/file` and it would end up trying to divide `path` by `to`.

If the input is a text string without a '/', like '123notafraction', adding the float cast will distort the result even more.

Are there any unit tests for this function at all ? Might it be an idea to add some and then to fix the function to only handle valid input and return `0` for invalid input ?

To illustrate:
Original: https://3v4l.org/ZepKd
This version: https://3v4l.org/rugAO

Let me know if I can help.

adamsilverstein commented:
> @adamsilverstein 👋🏻 Hope life is treating you well!

Hi @jrfnl :) thanks for the feedback, hope things are well with you!

You make good points, this function is a mess. 

>Are there any unit tests for this function at all ? Might it be an idea to add some and then to fix the function to only handle valid input and return 0 for invalid input ?

Tests would be helpful, I agree. I was trying to fix the PHP fatal error for images that are missing meta (and call this function with '0/0') without doing much else, but agree a more complete fix would be better.

Since this a bug fix we can continue to work on it, I'll add some more logic to enforce that (int)/(int) is what is passed, catch division by 0 etc, and add some tests that prove this out. returning 0 for any issues makes sense to me.

> Let me know if I can help.

I'll pick this up some time tomorrow US Mountain time as I'd like to get it in soon, if you have time and add something before then that would be fantastic.

jrfnl commented:
> I'll pick this up some time tomorrow US Mountain time as I'd like to get it in soon, if you have time and add something before then that would be fantastic.

Feel free to ping me on Slack when you start working on it and I'll let you know whether I'm available.

adamsilverstein commented:
> Feel free to ping me on Slack when you start working on it and I'll let you know whether I'm available.

Picking this up again, I'll ping you for review when ready!

adamsilverstein committed:
Improve logic in wp_exif_frac2dec and add tests

adamsilverstein committed:
Phpcs

adamsilverstein committed:
Fix zero check for PHP 7.

adamsilverstein committed:
Cleanup

adamsilverstein committed:
refine doc block

adamsilverstein committed:
Add test with int return

jrfnl reviewed (commented):
👋🏻 Hi Adam, had a quick look through and left some suggestions inline. Let me know what you think and feel free to ping me if you have questions.

adamsilverstein commented on src/wp-admin/includes/image.php:
```diff
@@ -647,18 +647,27 @@ function wp_generate_attachment_metadata( $attachment_id, $file ) {
-	if ( false === strpos( $str, '/' ) ) {
-		return $str;
+	// Fractions must contain a single `/`.
+	if ( false === strpos( $str, '/' ) || substr_count( $str, '/' ) > 1 ) {
```
yes, nice!

adamsilverstein committed:
Update src/wp-admin/includes/image.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

adamsilverstein committed:
Merge branch 'trunk' into ticket/54385

adamsilverstein committed:
Merge branch 'ticket/54385' of github.com:adamsilverstein/wordpress-develop-fork into ticket/54385

adamsilverstein commented on src/wp-admin/includes/image.php:
```diff
@@ -647,18 +647,27 @@ function wp_generate_attachment_metadata( $attachment_id, $file ) {
+	}
+
+	// The denominator must not be empty (or 0).
+	if ( empty( $denominator ) ) {
```
ill try that again after the regex change.

adamsilverstein commented on src/wp-admin/includes/image.php:
```diff
@@ -647,18 +647,27 @@ function wp_generate_attachment_metadata( $attachment_id, $file ) {
-		return $numerator / $denominator;
+
+	// Both the numerator and the denominator must be numbers.
+	if ( ! is_numeric( $numerator ) || ! is_numeric( $denominator ) ) {
```
> Question: could the numbers ever be floating point numbers ? Or will they always be integers ?

Great question. I dug into this a bit further. I reviewed a basket of images, some older some newer 

Core uses this function three times, passing data extracted from the image's Exif data by `exif_read_data` (http://exif.org/Exif2-2.PDF).
Once passing FNumber - something like `4.8` or `1.8` (once 'undef') in my testing - we round this and store the results in meta as `aperture`. I don't understand why we do that.
Once passing FocalLength - strings like `37 mm` or `6.0 mm (35 mm equivalent: 37.0 mm)` - we store the returned value in meta as focal_length after casting as a string
Once passing ExposureTime - in my test images this has a value like `1/1808` or `1/350` - should work as expected - we store the results in meta as shutter_speed after casting to a string.

Based on this usage, I'm thinking we should revert the 0 returns - so I'm going to modify the function as follows:

* When the passed value isn't a fraction, return the originally passed strings, thats what these calls all expect the function to do.
* Update doc block to indicate this is the expected behavior
* cast the result before rounding to float
* adjust tests to match, add test cases that match expected data from exploration above

adamsilverstein replied:
> Little known secret: is_numeric() is not always precise enough as it allows for float scientific annotations too.

Do you mean something like "4.321768×10^3" or "1.8281E-9"? and would 4.321768×10^3/100000 throw an error or work? I'm going to check. if it works, I feel like is_numeric is accurate here, the number can be a float or integer and the division will work.

adamsilverstein committed:
Update tests/phpunit/tests/image/functions.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

adamsilverstein commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -700,4 +700,94 @@ public function test_pdf_preview_doesnt_overwrite_existing_jpeg() {
+	 *
+	 * @ticket 54385
+	 * @dataProvider data_test_wp_exif_frac2dec
+	 */
```
👍🏼

adamsilverstein replied:
added

adamsilverstein committed:
Update tests/phpunit/tests/image/functions.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

adamsilverstein commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -700,4 +700,94 @@ public function test_pdf_preview_doesnt_overwrite_existing_jpeg() {
+	 *
+	 */
+	public function data_test_wp_exif_frac2dec() {
+		global $wpdb;
```
i guess; removed.

jrfnl replied:
Still there in the latest version...

adamsilverstein replied:
Removed!!!

adamsilverstein committed:
Update tests/phpunit/tests/image/functions.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

adamsilverstein commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -700,4 +700,94 @@ public function test_pdf_preview_doesnt_overwrite_existing_jpeg() {
+	 * Data provider for testing `wp_exif_frac2dec()`.
+	 *
+	 * @return array {
+	 *     Arguments for testing `wp_exif_frac2dec()`.
+	 *
+	 *     @type string|null        $fraction       The fraction string to run.
+	 *     @type int|float       $expected    The resulting expected value.
```
right, that is nicer because the tests each get a name as well

adamsilverstein committed:
Merge branch 'ticket/54385' of github.com:adamsilverstein/wordpress-develop-fork into ticket/54385

adamsilverstein committed:
merge

adamsilverstein committed:
Update src/wp-admin/includes/image.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

adamsilverstein committed:
Merge branch 'ticket/54385' of github.com:adamsilverstein/wordpress-develop-fork into ticket/54385

adamsilverstein commented on src/wp-admin/includes/image.php:
```diff
@@ -647,18 +647,27 @@ function wp_generate_attachment_metadata( $attachment_id, $file ) {
+	}
+
+	// The denominator must not be empty (or 0).
+	if ( $denominator === 0 ) {
```
testing in PHP8, I see `list( $numerator, $denominator ) = explode( '/', $str );` returning an empty denominator (not 0 but an empty string) for the test case '100/0'. Not certain why, but feels safer to use empty test here anyway.

jrfnl replied:
I cannot reproduce what you are saying: https://3v4l.org/llsu0

adamsilverstein replied:
ooh, that eval tool is cool!

I see what was happening now, the denominator is "0" in this case: that is a string of the number 0. 

In this case `0 ===` returns false, and still messes up the division. `empty()` returns true and original value is returned.

adamsilverstein replied:
since i restored is_numeric, "0" wouldn't make it here, will try restoring

adamsilverstein committed:
Cleanup after review

adamsilverstein committed:
phpcbf

adamsilverstein committed:
cleanup

adamsilverstein committed:
Merge branch 'trunk' into ticket/54385

jrfnl commented on src/wp-admin/includes/image.php:
```diff
@@ -647,18 +647,27 @@ function wp_generate_attachment_metadata( $attachment_id, $file ) {
 function wp_exif_frac2dec( $str ) {
-	if ( false === strpos( $str, '/' ) ) {
+	// Fractions must contain a single `/`.
+	if ( substr_count( $str, '/' ) !== 1 ) {
```
What about if someone passed a non-string ? Would it be an idea to use still slightly more defensive coding ?
```suggestion
	if ( ! is_scalar( $str ) || substr_count( $str, '/' ) !== 1 ) {
```

Suggestion: let's also add a couple of test cases with non-scalar input, like `null` and `array( '22/4' )`.
These would fail with a PHP error without the above adjustment.
(`null` would lead to a "passing null to non-nullable" deprecation notice in PHP 8.1 and `array()` would lead to a warning in PHP 7.x and a fatal error in PHP 8.x)

adamsilverstein replied:
great suggestion, will do!

jrfnl commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -700,4 +700,106 @@ public function test_pdf_preview_doesnt_overwrite_existing_jpeg() {
+				'fraction' => '0/abc',
+				'expect'   => '0/abc',
+			),
+			'non fraction returns original value' => array(
+				'fraction' => '0.0',
+				'expect'   => '0.0',
+			),
+			'non fraction returns original value' => array(
+				'fraction' => '010',
+				'expect'   => '010',
+			),
+			'non fraction returns original value' => array(
+				'fraction' => 10.123,
+				'expect'   => 10.123,
+			),
```
Only the last of these test cases will currently run as you're using the exact same (duplicate) array key.

jrfnl commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -700,4 +700,106 @@ public function test_pdf_preview_doesnt_overwrite_existing_jpeg() {
+				'fraction' => '4/2',
+				'expect'   => 2,
+			),
+			'non fraction returns original value' => array(
+				'fraction' => '0',
+				'expect'   => '0',
+			),
+			'division by zero is prevented'       => array(
+				'fraction' => '100/0',
+				'expect'   => '100/0',
+			),
+			'text is prevented'                   => array(
+				'fraction' => 'path/to/file',
+				'expect'   => 'path/to/file',
+			),
+			'text is prevented'                   => array(
+				'fraction' => '123notafraction',
+				'expect'   => '123notafraction',
+			),
```
More duplicate array keys

adamsilverstein replied:
Didn't realize that would be an issue, I've become a golang programmer :)

jrfnl reviewed (commented):
Left some more notes

adamsilverstein committed:
improve test cases

adamsilverstein committed:
Update src/wp-admin/includes/image.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

adamsilverstein committed:
Merge branch 'ticket/54385' of github.com:adamsilverstein/wordpress-develop-fork into ticket/54385

adamsilverstein committed:
Remove global $wpdb;

adamsilverstein commented:
> Still there in the latest version...
removed in https://github.com/WordPress/wordpress-develop/pull/1841/commits/be0030df033a7d82c023fa88bbae7f4441f80a49

adamsilverstein committed:
Merge branch 'trunk' into ticket/54385

adamsilverstein committed:
phpcs!

adamsilverstein committed:
Improve docblock.

adamsilverstein committed:
Cast return from wp_exif_frac2dec as float before applying round.

adamsilverstein committed:
improve denominator 0 check.

jrfnl committed:
wp_exif_frac2dec(): suggested changes

These are the changes I'd like to suggest.

1. Maintain the return type for the function as `int|float` as was previously already documented.
2. Return integer `0` for all forms of invalid input, but be tolerant of integers/floats and fully numeric integer/float strings passed to the function.
3. In the places where the function is used, largely maintain the previous behaviour by setting the original value as the default and only overwriting it with the output of `wp_exif_frac2dec()` if the output is likely to be more relevant.
    Note: it would be good to add tests for the `wp_read_image_metadata()` function as well to verify and safeguard this.
4. Order the test cases to follow the logic of the function and update them for the proposed changes.
    Includes making sure that each test case has a unique and descriptive name for the test case.

jrfnl committed:
wp_read_image_metadata(): minor reduction in defensive coding

jrfnl committed:
Tests_Image_Meta: add $message parameter to tests with multiple assertions

adamsilverstein commented:
Thanks for the updates @jrfnl and your feedback on the approach. Now that you have pointed out `Tests_Image_Meta` I think it would make sense to add at least one of the two test images from the ticket (I'll double check that the issue is the same with both images). I'll work on that.

adamsilverstein committed:
Add sugarloaf-mountain.jpg image and test.

adamsilverstein committed:
Merge branch 'trunk' into ticket/54385

adamsilverstein committed:
document test

adamsilverstein committed:
improve test doc block

peterwilsoncc commented on src/wp-admin/includes/image.php:
```diff
@@ -646,19 +646,40 @@ function wp_generate_attachment_metadata( $attachment_id, $file ) {
+	}
+
+	// The denominator must not be zero.
+	if ( 0 == $denominator ) { // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison -- Deliberate loose comparison.
```
```suggestion
	// phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison -- Deliberate loose comparison.
	if ( 0 == $denominator ) {
```

`phpcs:ignore` ignores [both the current and next line](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-parts-of-a-file) so putting it on a line above avoids accidentally ignoring more than is intended.

An expanded comment explaining that this can't be converted to a falsey check due to the possibility of '0.0' would be good too https://3v4l.org/LiF4J

jrfnl replied:
@peterwilsoncc The "ignoring the current and next line" only applies to lines which only contain the ignore annotation, not to trailing comments.

The reason I put it as a trailing comment instead of stand-alone is that there is already a comment above the line.
Inserting the ignore annotation between them would divorce the comment from the code it applies to and _might_ tempt someone to then convert the two comment lines to a multi-line `/* */` comments syntax, which would break the ignore annotation.

Expanding the comment wouldn't hurt, but there is a test covering this exact case `'0.0'`, so anyone tempted to change this should be alerted soon enough that that's not a good idea and why due to that specific test failing. (it's the reason I put the test in).

hellofromtonya commented:
Committed via changeset https://core.trac.wordpress.org/changeset/52269.
